File: app/src/main/java/com/nimtego/plectrum/domain/interactor/SongInteractor.java
package com.nimtego.plectrum.domain.interactor;

import com.nimtego.plectrum.presentation.main.model.SongModelK;

import java.util.List;

import dagger.internal.Preconditions;
import io.reactivex.Observable;

public class SongInteractor extends BaseInteractor<List<SongModelK>, SongInteractor.Params> {

    public SongInteractor() {
    }

    @Override
    protected Observable<List<SongModelK>> buildUseCaseObservable(Params params) {
        Preconditions.checkNotNull(params);
        return repository.songs(params.request);
    }

    public static final class Params {

        private final String request;

        private Params(String request) {
            this.request = request;
        }

        public static Params forRequest(String request) {
            return new Params(request);
        }
    }

}

File: app/src/main/java/com/nimtego/plectrum/domain/interactor/SongInteractor.kt
package com.nimtego.plectrum.domain.interactor

import com.nimtego.plectrum.presentation.mvp.view_model.music.SongModel
import io.reactivex.Observable


class SongInteractor : BaseInteractor<List<SongModel>, SongInteractor.Params>() {

    override fun buildUseCaseObservable(params: Params): Observable<List<SongModel>> {
        return repository.songs(params.request)
    }

    class Params private constructor(val request: String) {
        companion object {
            fun forRequest(request: String): Params {
                return Params(request)
            }
        }
    }

}